python/ray/train/_internal/storage.py

```diff
@@ -0,0 +1,236 @@
+
+    def _create_validation_file(self):
+        """On the creation of a storage context, create a validation file at the
+        storage path to verify that the storage path can be written to.
+        This validation file is also used to check whether the storage path is
+        accessible by all nodes in the cluster."""
+        valid_file = os.path.join(self.experiment_fs_path, ".validate_storage_marker")
+        self.storage_filesystem.create_dir(self.experiment_fs_path)
+        with self.storage_filesystem.open_output_stream(valid_file):
+            pass
+
+    def _check_validation_file(self):
+        """Checks that the validation file exists at the storage path."""
+        valid_file = os.path.join(self.experiment_fs_path, ".validate_storage_marker")
+        valid = self.storage_filesystem.get_file_info([valid_file])[0]
+        if valid.type == pyarrow.fs.FileType.NotFound:
+            raise RuntimeError(
+                f"Unable to set up cluster storage at storage_path={self.storage_path}"
+                "\nCheck that all nodes in the cluster have read/write access "
+                "to the configured storage path."
+            )
+
+    @property
+    def experiment_fs_path(self) -> str:
+        """The path on the `storage_filesystem` to the experiment directory."""
+        return os.path.join(self.storage_fs_path, self.experiment_dir_name)
+
+    @property
+    def experiment_cache_path(self) -> str:
+        """The local filesystem path to the experiment directory.
+
+        This local "cache" path refers to location where files are dumped before
+        syncing them to the `storage_path` on the `storage_filesystem`.
+        """
+        return os.path.join(self.storage_cache_path, self.experiment_dir_name)
+
+    @property
+    def trial_fs_path(self) -> str:
+        """The trial directory path on the `storage_filesystem`.
+
+        Raises a ValueError if `trial_dir_name` is not set beforehand.
+        """
+        if self.trial_dir_name is None:
+            raise RuntimeError(
+                "Should not access `trial_fs_path` without setting `trial_dir_name`"
+            )
+        return os.path.join(self.experiment_fs_path, self.trial_dir_name)
+
+    @property
+    def checkpoint_fs_path(self) -> str:
+        """The trial directory path on the `storage_filesystem`.
+
+        Raises a ValueError if `current_checkpoint_id` is not set beforehand.
+        """
+        if self.current_checkpoint_id is None:
+            raise RuntimeError(
+                "Should not access `checkpoint_fs_path` without setting "
+                "`current_checkpoint_id`"
+            )
+        checkpoint_dir_name = TrainableUtil._make_checkpoint_dir_name(
+            self.current_checkpoint_id
+        )
+        return os.path.join(self.trial_fs_path, checkpoint_dir_name)
+
+
+_storage_context: Optional[StorageContext] = None
+
+
+def init_shared_storage_context(storage_context: StorageContext):
```

Comment:
```suggestion
def init_trial_storage_context(storage_context: StorageContext):
```

Reply:
I'm actually thinking of keeping this name generic, since we may want to create a shared copy in the train workers as well. When we move to a unified implementation of `train.report`, the logic would involve getting this storage context somehow.

`trial_storage_context` may also be a bit misleading, since the actual `trial` objects live on the driver.